File: question_9/py/solution.py
#!/usr/bin/env python3
"""Solutuon to vagina number using recursion
"""


def vagina_number_checker(number: int, f: str = None, s: str = None) -> bool:
    """Function that checks if a number is a vagina number
    Params:
        - number: <int> number to check if it's a valid vagina
        - f: <str> first digit in number assuming number is only 3 digit
        - s: <str> second digit in number assuming number is only 3 digit

    Return:
        - True if number is a valid vagina otherwise false
    """
    # assume all numbmers have a length of 3
    N = str(number)
    if len(N) != 3:
        return False
    if f and s:
        return True if int(f) * int(s) == int(N[2]) else False

    return vagina_number_checker(int(N), N[0], N[1])


def main(N: int, first_check: bool = None, second_check: bool = None,
         count: int = 0, res: int = None) -> int:
    """Entry point to solution
    params:
        - N: <int> input

    Return:
        - Following the Constraints 100≤N≤919 returns the smallest vagina
        - number that is greater than of equal to N
    """
    if 100 <= N and N <= 919:
        # continue
        if first_check and second_check:
            return res

        if not vagina_number_checker(N) and count == 0:
            return main(N + 1, None, None, count, res)

        if not vagina_number_checker(N) and count > 0:
            return main(N + 1, first_check, None, count, res)

        if vagina_number_checker(N) and count > 0:
            return main(N + 1, True, True, count + 1, res)

        if vagina_number_checker(N):
            # first check if N is vagina equals true
            return main(N + 1, True, None, count + 1, N)
    return False


print(main(320))
print(main(144))
print(main(516))
print(main(122))
print(main(100))

"""
main(224)
print("\n")
main(287)
main(473)
main(112)
"""

File: question_9/c/solution.c
#include "main.h"


/**
* vagina_number_checker - function that checks if a number is a valid vagina
* number using recursion
*
* @number: <int> number to perfrom check on
* @first: <string> reference to the first character in number
* @second: <string> reference to the second character in number
*
* Return: True if number is valid vagina number otherwise False
*/
bool vagina_number_checker(int number, const char *first, const char *second)
{
	char str[4];

	snprintf(str, sizeof(str), "%d", number);

	size_t length = strlen(str);

	if (length != 3)
	{
		return (false);
	};
	char f[2] = { str[0], '\0' }, s[2] = { str[1], '\0' };
	char last[2] = { str[2], '\0' };

	if (first && second)
	{
		return (atoi(first) * atoi(second) == atoi(last) ? true : false);
	}
	return (vagina_number_checker(number, f, s));
}


/**
* vagina_main - entry point
*
* @N: <int> input number to find the smallest
* @first_check: <bool> first reference to when a vagina number exits
* @second_check: <bool> second reference to when a vagina number exists
* @count: <int> counter once a vagina number has been found
* @result: <int> output
*
* Return: smallest vagina number that is greater than or equal to N
 */
int vagina_main(int N, bool first_check, bool second_check, int count,
		int result)
{
	if (N >= 100 && N <= 919)
	{
		if (first_check && second_check)
		{
			return (result);
		};

		if (!vagina_number_checker(N, NULL, NULL) && count == 0)
		{
			return (vagina_main(N + 1, NULL, NULL, count, result));
		};

		if (!vagina_number_checker(N, NULL, NULL) && count > 0)
		{
			return (vagina_main(N + 1, first_check, NULL, count, result));
		};

		if (vagina_number_checker(N, NULL, NULL) && count > 0)
		{
			return (vagina_main(N + 1, true, true, count + 1, result));
		};

		if (vagina_number_checker(N, NULL, NULL))
		{
			return (vagina_main(N + 1, true, NULL, count + 1, N));
		};
	};
	return (0);
}


/**
* main - entry point
*
* Return: 0
 */
int main(void)
{
	int numbers[] = {320, 144, 516, 122, 100};

	for (int i = 0; i < sizeof(numbers) / sizeof(numbers[0]); i++)
	{
		printf("%d\n", vagina_main(numbers[i], NULL, NULL, '\0', '\0'));
	};
	return (0);
}

